### DISCUSSION TITLE
How to enable experimental features on Gen1 Xiaomi?

### DISCUSSION THREAD

**Discussion by fokcuk** [rand256/valetudo#436] [Q&A]
Hi everyone.
After finally installing Valetudo on my Gen1 vac, I found that I had issues with maps. So I went to RE =)
I had some issues with 4028 (map was rotated for the half of the house), went back to 4018.
But I cannot get any experimental features to work =( Issues I have so far:
- I do not see any options for the bin empty feature. Probably in the missing lab section? Or do I have to enable it separately somehow?
- Valetudo does not send map data to HA. I can control it from HA, get its state etc, but nothing else. No other entities found.
- How do I enable debug logging?

**Comment by rand256** [comment]
> I do not see any options for the bin empty feature. Probably in the missing lab section? Or do I have to enable it separately somehow?

What is the bin empty feature exactly?
The device adds `last_bin_out` and `last_bin_full` fields to json object published to `.../attributes` topic of mqtt. That's all it does regarding the bin.

> Valetudo does not send map data to HA. 

It doesn't send anything to HA specifically. It can just publish the map on mqtt broker, it's up to HA to look up it there. Check that it is enabled on mqtt settings page. Also it should be noted that it publishes the original map data generated by firmware, it is in binary format and not parsed or converted to json.

> How do I enable debug logging?

There's none.
The only log that exists is what you can see at Settings -> Info tab in WebUI or in /var/log/upstart/valetudo.log

**Reply by fokcuk** [reply]
Right... just understand that I dont understand =)
This is what I see in the HA log:
1622211838: New client connected from 192.168.1.123 as mqttjs_a34a2e21 (p2, c1, k60).
1622212353: Socket error on client mqttjs_a34a2e21, disconnecting.
1622212412: New connection from 192.168.1.123 on port 1883.
1622212412: New client connected from 192.168.1.123 as mqttjs_76182c24 (p2, c1, k60).
1622212864: Socket error on client mqttjs_76182c24, disconnecting.
1622212890: New connection from 192.168.1.123 on port 1883.
1622212890: New client connected from 192.168.1.123 as mqttjs_eed07a7e (p2, c1, k60).

This is what I get in the HA MQTT entity:
rockrobo' (vacuum.rockrobo)
MQTT discovery data:
Topic: homeassistant/vacuum/valetudo_rockrobo/config
Payload
Subscribed topics:
valetudo/rockrobo/attributes
10 most recently received message(s)
valetudo/rockrobo/state
10 most recently received message(s)

Do I need to subscribe to more topics? Thought it would just find them as robot should be pushing them out to the HA.

OK. Just opened up the MTTQ message in HA:
bin_in_time: 266
last_bin_out: 1622260321040
last_bin_full: -1
Now need to find that these numbers mean =)

Now, map data...It is checked in the GUI of Robo. Now, in the below topic, there are no other topics listed that are relevant to the maps:
Topic: homeassistant/vacuum/valetudo_rockrobo/config
Payload
name: rockrobo
unique_id: rockrobo
schema: state
device:
  manufacturer: Roborock
  model: rockrobo.vacuum.v1
  name: rockrobo
  identifiers:
    - rockrobo
  sw_version: 0.10.7
supported_features:
  - start
  - pause
  - stop
  - return_home
  - battery
  - status
  - locate
  - clean_spot
  - fan_speed
  - send_command
command_topic: valetudo/rockrobo/command
state_topic: valetudo/rockrobo/state
set_fan_speed_topic: valetudo/rockrobo/set_fan_speed
fan_speed_list:
  - whisper
  - min
  - medium
  - high
  - max
send_command_topic: valetudo/rockrobo/custom_command
json_attributes_topic: valetudo/rockrobo/attributes
platform: mqtt

**Reply by rand256** [reply]
>bin_in_time: 266
> last_bin_out: 1622260321040
> last_bin_full: -1
> Now need to find that these numbers mean

It's in the [wiki](https://github.com/rand256/valetudo/wiki/MQTT-API#Topic_valetudorockroboattributes).

> Now, map data...

In which format do you want it do be?
Probably your HA can only process JSON while the map valetudo publishes is binary packed.

If you want to have a map as a picture, see valetudo-mapper app.

**Reply by fokcuk** [reply]
I was referring to the map-data-hass-hack part. Suppose to send it to Home Assistant as part of the discovery as a PNG image
https://valetudo.cloud/pages/integrations/mqtt.html#rawmapdatawithhomeassistanthackmap-data-hass-hack

**Reply by rand256** [reply]
It does not exist in RE. Also I have no idea what it actually is and since I don't use any HA then probably will neither know it in the future. Feel free to implement missing parts in RE and make a PR if you wish.

**Reply by fokcuk** [reply]
wish I could =)
Just trying to find an easy way out. I tried using the image from http://192.168.1.123/api/simple_map, but it does not work inside HA, as its not really an image in HA's eyes.
I would love to get map view integrated, but everything is new to me - have pulled your Docker container into HASSOS, but have no idea what to do next =) A lot to learn
